multiproof.go

```diff
@@ -69,27 +81,32 @@ func CreateMultiProof(transcript *common.Transcript, ipaConf *ipa.IPAConfig, Cs
 	transcript.AppendPoint(&D, "D")
 	t := transcript.ChallengeScalar("t")
 
-	// Compute h(X) = g_1(X)
-	h_x := make([]fr.Element, common.POLY_DEGREE)
-
-	den_inv := make([]fr.Element, num_queries)
-	for i := 0; i < num_queries; i++ {
-		var z = domainToFr(zs[i])
-		den_inv[i].Sub(&t, &z)
+	// Calculate the denominator inverses only for referenced evaluation points.
+	den_inv := make([]fr.Element, 0, common.POLY_DEGREE)
+	for z, f := range groupedFs {
+		if len(f) == 0 {
+			continue
+		}
+		var z = domainToFr(uint8(z))
+		var den fr.Element
+		den.Sub(&t, &z)
+		den_inv = append(den_inv, den)
 	}
 	den_inv = fr.BatchInvert(den_inv)
-	for i := 0; i < num_queries; i++ {
-		r := powers_of_r[i]
-		f := fs[i]
 
+	// Compute h(X) = g_1(X)
+	h_x := make([]fr.Element, common.POLY_DEGREE)
+	denInvIdx := 0
+	for _, f := range groupedFs {
```

Comment:
Compute the final `g_1(x)` using the grouped polynomials, and the computed denominator inverses.